### PR TITLE
Minify js and css files and add them to build folder #5487

### LINKED ISSUE #5487: Publish minified files to pdfjs-dist npm package
The npm `pdfjs-dist` package should containt minified file as well. Most front-end projects publishing on npm does it.

### PR DESCRIPTION
**Pull request by meltuhamy**
- follows on from #5593
- use npm for dependency on closure compiler and clean-css
- remove unneeded commas

Fixes #5487.

### CODE REVIEW

File: make.js [hunk 1]
```diff
@@ -168,20 +168,20 @@ target.components = function() {
     'web/images/annotation-*.svg',
     'web/images/loading-icon.gif',
     'web/images/shadow.png',
-    'web/images/texture.png',
```
Comment (yurydelendik) [inline make.js]: unrelated changes (here and below), please revert.

**Comment by meltuhamy** [comment]
Hi @yurydelendik I've removed unnecessary changes as per your suggestion, and also fixed an issue which was causing the minified viewer not to display properly.  

A `minified` folder is now produced that contains all the needed minified files and also a fixed `viewer.html` that shows the preview pdf without errors.

**Comment by mzgol** [comment]
Could minified files be put in the same directory as the built un-minified ones? Some tools assume that will be the case and they just swap `.js` to `.min.js`.

**Comment by meltuhamy** [comment]
@mzgol thanks - my latest push now does this - however, there is a subtle difference: `pdf_viewer.min.js` *includes* `pdf.js`, while `pdf_viewer.js` *does not include* `pdf.js`.

The `dist` directory structure looks like this: 

```
dist
|   bower.json
|   LICENSE
|   package.json
|   README.md
|
+---build
|       pdf.combined.js
|       pdf.combined.min.js
|       pdf.js
|       pdf.min.js
|       pdf.worker.js
|       pdf.worker.min.js
|
+---cmaps
|       *.bcmap
|       LICENSE
|
\---web
    |   compatibility.js
    |   compatibility.min.js
    |   pdf_viewer.css
    |   pdf_viewer.js
    |   pdf_viewer.min.js
    |   viewer.min.css
    |
    \---images
            *.svg
            *.png

```

File: make.js [hunk 2]
```diff
@@ -637,47 +645,78 @@ function cleanupCSSSource(file) {
...
 target.minified = function() {
-  var compilerPath = process.env['CLOSURE_COMPILER'];
-  if (!compilerPath) {
-    echo('### Closure Compiler is not set. Specify CLOSURE_COMPILER variable');
+  try {
+    var compilerPackagePath = require.resolve('google-closure-compiler');
+    var compilerDirPath = path.dirname(compilerPackagePath);
+    var compilerPath = path.join(compilerDirPath, 'compiler.jar');
+  } catch (e) {
+    console.log('google-closure-compiler is not installed. Run "npm install" ' +
```
Comment (Snuffleupagus) [inline make.js]: Is there reason that you're using `console.log` instead of `echo` here (and below)?

Comment (meltuhamy) [inline make.js]: Not really, I'm just used to console.log - happy to change to echo for consistency.

**Comment by meltuhamy** [comment]
Code changed and ready for review again :)

**Comment by mzgol** [comment]
> @mzgol thanks - my latest push now does this - however, there is a subtle difference: pdf_viewer.min.js includes pdf.js, while pdf_viewer.js does not include pdf.js.

Why is that necessary? Ideally `FILENAME.js` and `FILENAME.min.js` should differ only in that the latter is minified...

**Comment by meltuhamy** [comment]
I agree but that would mean having 1 extra script tag so potentially 3 script requests would be made for the viewer to load. Also, there is no use having the viewer code without pdfjs loaded - ie the two files will always be required together. If a consumer needs pdfjs separately, that is also provided on its own in the build folder. 

The alternative is to provide the viewer code minified separately, and a combined viewer+pdfjs file in another file if anyone wishes to use that. 

Let me know what you think.

**Comment by mzgol** [comment]
> The alternative is to provide the viewer code minified separately, and a combined viewer+pdfjs file in another file if anyone wishes to use that.

Note that I'm not a maintainer of this module so all I say is only my personal opinion and the maintainers might decide otherwise. So you might want to wait for an official opinion before changing anything.

That said, I think that would be great to have a combined viewer+pdfjs file and then its minified version. This would go in line with conventions other libraries take so it would reduce confusion in developers using pdf.js.

**Comment by yurydelendik** [comment]
Currently our dist, web and release packaging is a little bit messy, but I agree that we shall match convention @mzgol mentioned above. Let's think if can / cannot package minified files in some cases without original source

**Comment by meltuhamy** [comment]
@yurydelendik I've updated the branch, it should now follow the convention as suggested by @mgol 

The output now looks like the following:
```
+---minified
|   |   LICENSE
|   |   
|   +---build
|   |       pdf.combined.js
|   |       pdf.combined.min.js
|   |       pdf.js
|   |       pdf.min.js
|   |       pdf.worker.js
|   |       pdf.worker.min.js
|   |       
|   \---web
|       |   compatibility.js
|       |   compatibility.min.js
|       |   compressed.tracemonkey-pldi-09.pdf
|       |   debugger.js
|       |   l10n.js
|       |   pdf_viewer.min.js
|       |   viewer.css
|       |   viewer.html
|       |   viewer.js
|       |   viewer.min.css
|       |   viewer.min.js
|       |   
|       +---cmaps
|       |       *.bcmap
|       |     
|       +---images
|       |       *
|       |       
|       \---locale
|       |       *
```

**Comment by meltuhamy** [comment]
What are your thoughts on packaging `preview.html` with the rest of the files? It would be good to provide the whole viewer as a minified application.

**Comment by timvandermeij** [comment]
This patch needs a rebase after the UglifyJS patch that landed today. Additionally, please squash the commits into one afterwards (see https://github.com/mozilla/pdf.js/wiki/Squashing-Commits).

**Comment by meltuhamy** [comment]
Will update this, but what are your thoughts on packaging `index.html` with the minified files? This means there is a demo that uses the minified files and also allows people to consume the reader as a whole (Eg. By putting it into an iframe) a lot easier. 

Let me know your thoughts.

**Comment by timvandermeij** [comment]
We have `node make minified` to get the (minified) library and the viewer. The NPM package however should only contain the (minified) library. We can always add the viewer in a follow-up patch, but a while ago it was decided to not distribute the viewer via NPM because most people don't want to use the entire viewer (as it's specific to the Firefox use-case) and only make use of the component-based approach. On top of that, we ask developers to alter the default viewer to make it different from the viewer in Firefox, to not confuse users. If we ship the default viewer, this is less likely to happen.

File: make.js [hunk 3]
```diff
@@ -307,11 +309,22 @@ target.dist = function() {
...
     SINGLE_FILE_DIR + 'build/pdf.combined.js',
+    MINIFIED_DIR + 'build/pdf.min.js',
+    MINIFIED_DIR + 'build/pdf.worker.min.js',
+    MINIFIED_DIR + 'build/pdf.combined.min.js'
   ], DIST_DIR + 'build/');
 
   mkdir('-p', DIST_DIR + 'web/');
   cp('-R', [
     COMPONENTS_DIR + '*',
+    GENERIC_DIR + 'web/viewer.js',
```
Comment (Snuffleupagus) [inline make.js]: Please note that the *only* files that should be present in the `dist` are the ones listed here: https://github.com/mozilla/pdfjs-dist/tree/master/web.
Providing minified versions of those files should be OK, but `viewer.{js, css}` should *not* be included!

**Comment by mgol** [comment]
> The NPM package however should only contain the minified library.

I think it should contain both minified & unminified libraries. That's a pattern most libraries follows and it tremendously helps in debugging various issues. The only difference between the minified & unminified version should be the minification itself.

**Comment by timvandermeij** [comment]
@mgol Sorry, indeed I meant that the package should contain the minified and unminified library, just not the viewer files.

File: make.js (hunk 3 above)
Comment (meltuhamy) [inline make.js]: I think you're right, it's maybe not in the scope of this PR, but I also think it would be good to publish a minified viewer. 

The use case I have in mind is if we want to export a generic viewer that can be included "out of the box" in other projects. Perhaps this is not the scope of the pdf.js project (it could be extracted into its own project that uses pdf.js), but then perhaps the whole viewer should not be included in this repo in that case. I appreciate it's quite a big and perhaps unnecessary change but if we're saying that the viewer shouldn't be published then it might make sense to put the viewer into its own repo and not version it here. 

I guess my question is: what is the role of the viewer in this repo? If it is a product that is bring released then maybe there's no harm on releasing a minified version.

Comment (timvandermeij) [inline make.js]: The viewer is this repository is not intended to be used as-is by third-party solutions. Third-party viewers should always be different from the default viewer. Most people want a simple viewer, which they can achieve by using https://github.com/mozilla/pdf.js/tree/master/examples/components *or* by using parts of the default viewer. The default viewer is in this repository for two reasons: open-source development for the viewer in Firefox (and the Chrome extension) and third-party viewers might reuse parts of it.

It is really out of scope for this PR to include the viewer though.

Comment (meltuhamy) [inline make.js]: Cool, thanks to the response. I'll patch the pr with the required changes  and squash it. Thanks

**Comment by meltuhamy** [comment]
This is now done @timvandermeij

**Comment by meltuhamy** [comment]
This is no longer needed as uglify js has been brought in :)
